### PR TITLE
[OPIK-6028] [CI] perf: build Docker images once via GHCR, parallelize E2E tests

### PR DESCRIPTION
**Pull request by JetoPistola**
## Details

image

All 6 E2E test workflows (Python SDK, Python SDK Compat V1, TypeScript SDK, TypeScript SDK Compat V1, SDK Libraries, Opik Optimizer) previously built Docker images independently in every matrix entry. This meant up to 26 redundant Docker builds per PR.

This PR introduces a reusable `build_e2e_docker.yaml` workflow that builds once per workflow, pushes to GHCR with immutable SHA-based tags, and test runners pull pre-built images. A cleanup workflow (`cleanup_e2e_docker.yaml`) deletes stale images on PR close and via weekly Saturday sweep.

**Key changes:**

- **New reusable workflow** (`build_e2e_docker.yaml`): builds backend + python-backend (+ guardrails-backend when needed), pushes to `ghcr.io/comet-ml/opik/` with immutable `ci-e2e-pr<N>-<sha7>` tags
- **6 consumer workflows updated**: replaced inline `--build` with `docker pull` from GHCR + retag to `:latest`
- **docker-compose.yaml**: made `backend` and `guardrails-backend` `pull_policy` configurable via env vars (defaults to `always`, set to `never` in CI)
- **New cleanup workflow** (`cleanup_e2e_docker.yaml`): deletes all `ci-e2e-pr<N>-*` tags on PR close (immediate) + weekly Saturday sweep at 03:00 UTC (safety net, age-based with configurable `max_age_days`)

**Tag scheme** (immutable, unique per push):
- PR trigger: `ci-e2e-pr<N>-<sha7>` (e.g., `ci-e2e-pr777-abc123d`)
- PR trigger with guardrails: `ci-e2e-pr<N>-<sha7>-guardrails`
- Non-PR trigger (workflow_dispatch/schedule/main): `ci-e2e-<sha7>`

Each push creates a new unique tag — no overwrites, no orphaned untagged manifests, no race conditions. This is the industry standard for container CI.

| Metric | Before | After | Future (OPIK-6039) |
|---|---|---|---|
| Docker builds per PR | ~26 (per matrix runner) | ~6 (per workflow) | 2 (orchestrator) |
| Runner setup time | ~4m (build from scratch) | ~0.5–2.5m (pull + startup) | same |
| Flaky test re-run | ~10–12m (rebuild) | ~3–9m (pull only) | same |
| Image cleanup | N/A | On PR close + weekly Saturday | same |

**Concurrent workflow race**: When multiple workflows trigger on the same PR push, each builds independently with the same immutable tag (same SHA = same content = same tag). Docker registry pushes are atomic and idempotent — concurrent pushes of identical content are safe. But still wasteful (up to 6 identical builds instead of 2). Follow-up OPIK-6039 introduces a shared orchestrator to eliminate this.

**Measured CI times** (from PR #6764's run vs main baselines):
- TS SDK E2E runner setup: 4m20s → 1m23s (pull 27s + startup 56s)
- Python SDK E2E runner setup: 4m06s → 3m27s (pull 2m26s + startup 1m01s, 3 images w/ guardrails)
- Flaky re-run: skips build job entirely — images already in GHCR

**Cleanup workflow**:
- **PR close**: deletes all `ci-e2e-pr<N>-*` versions for that PR (catches all historical pushes)
- **Weekly Saturday sweep**: deletes all `ci-e2e-*` versions older than `max_age_days` (default 7, configurable via `workflow_dispatch` with `type: number` validation)
- No untagged cleanup needed — immutable tags don't produce orphaned untagged manifests
- Uses `GITHUB_TOKEN` with `packages: write` (no PAT needed)

> Re-creation of PR #6764 which was closed for cleanup workflow testing. Same commits, clean branch from main.

## Change checklist
- [x] User facing
- [x] Documentation update

## Issues

- OPIK-6028

## AI-WATERMARK

AI-WATERMARK: yes

- Tools: Claude Code
- Model(s): Claude Opus 4.6
- Scope: full implementation
- Human verification: code review + CI validation + manual workflow dispatch testing

### CI end-to-end validation — all 6 E2E workflows green

All 6 E2E workflows ran on this PR (commit `7ad8de5`). The workflows themselves are the test:
1. `build_e2e_docker.yaml` builds and pushes images to GHCR with immutable `ci-e2e-pr7117-<sha7>` tags
2. Each matrix runner pulls images from GHCR, retags to `:latest`, starts Opik server with `pull_policy: never`
3. Existing test suites run against the pre-built images

**CI run links (all green):**

| Workflow | Run | Status |
|---|---|---|
| Python SDK E2E (5 Python versions) | [27693149735](https://github.com/comet-ml/opik/actions/runs/27693149735) | ✅ passed |
| Python SDK Compat V1.x (5 Python versions) | [27693147220](https://github.com/comet-ml/opik/actions/runs/27693147220) | ✅ passed |
| TypeScript SDK E2E (3 Node versions) | [27693149683](https://github.com/comet-ml/opik/actions/runs/27693149683) | ✅ passed |
| TypeScript SDK Compat V1.x (5 Node versions) | [27693149650](https://github.com/comet-ml/opik/actions/runs/27693149650) | ✅ passed |
| SDK E2E Libraries Integration | [27693149808](https://github.com/comet-ml/opik/actions/runs/27693149808) | ✅ passed |
| Opik Optimizer E2E (5 Python versions + Smoke) | [27693149727](https://github.com/comet-ml/opik/actions/runs/27693149727) | ✅ passed |

**Results summary:**
- **6 build jobs** (one per workflow): all passed — GHCR push with immutable SHA tags confirmed working
- **22+ GHCR pull runners** across all workflows: all passed — pull + retag + `pull_policy: never` working
- **TS SDK E2E**: passed on re-run (first run hit external LLM API timeouts — flaky, unrelated)
- **Python SDK E2E**: 4/5 passed first try; Python 3.11 had a flaky timing assertion, passed on re-run

### Cleanup workflow validation — both trigger paths tested

Both cleanup trigger paths have been tested across PR #6764 (original, closed for testing) and PR #7117 (this PR):

| Test | Trigger | Run | Result |
|---|---|---|---|
| PR #6764 close | `pull_request: [closed]` | [26108094902](https://github.com/comet-ml/opik/actions/runs/26108094902) | ✅ Triggered on close, deleted 3/6 images before being externally cancelled |
| PR #6764 stale sweep | `workflow_dispatch` (`max_age_days=0`) | [26099058327](https://github.com/comet-ml/opik/actions/runs/26099058327) | ✅ Deleted 5 images for closed PR #6764, 0 failures |
| PR #7117 stale sweep | `workflow_dispatch` (`max_age_days=0`) | [27694643974](https://github.com/comet-ml/opik/actions/runs/27694643974) | ✅ Deleted 30 images across 3 repos, 0 failures |

**PR-close trigger (run 26108094902)**: When PR #6764 was closed, the `pull_request: [closed]` event fired and the `cleanup-pr` job correctly extracted PR number 6764 and began deleting `ci-e2e-6764*` versions. It successfully deleted `opik-backend:ci-e2e-6764`, `opik-backend:ci-e2e-6764-guardrails`, and `opik-python-backend:ci-e2e-6764` before being cancelled (the remaining images were cleaned up by subsequent `workflow_dispatch` runs).

**Weekly sweep / `workflow_dispatch` (run 27694643974)**: Triggered manually on this PR's branch with `max_age_days=0`. Scanned all 3 image repos (opik-backend, opik-python-backend, opik-guardrails-backend), found and deleted 12 + 12 + 6 = 30 ci-e2e-* versions. `cleanup-pr` job correctly skipped (not a PR close event). Confirms age-based deletion logic, pagination across ~31K package versions, and `GITHUB_TOKEN` with `packages: write`.

### What can only be tested after merge

- `schedule` cron trigger (Saturday 03:00 UTC) — requires the workflow definition to exist on `main`

## Documentation

N/A

### CODE REVIEW

File: .github/workflows/build_e2e_docker.yaml [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+    steps:
+      - name: Set image tag
+        id: tag
+        run: |
+          if [ -n "${{ github.event.pull_request.number }}" ]; then
+            TAG="ci-e2e-${{ github.event.pull_request.number }}${{ inputs.include_guardrails == true && '-guardrails' || '' }}"
+          else
+            TAG="ci-e2e-${{ github.sha }}${{ inputs.include_guardrails == true && '-guardrails' || '' }}"
+          fi
+          echo "tag=$TAG" >> $GITHUB_OUTPUT
```
[the remark this answers is not in the document]
Comment (JetoPistola) [reply .github/workflows/build_e2e_docker.yaml]: Skipping — `workflow_call` reusable workflows inherit the caller's full `github` context, including `github.event.pull_request.number`. When a consumer workflow triggers on `pull_request`, the called `build_e2e_docker.yaml` correctly sees the PR number. For non-PR triggers (workflow_dispatch, schedule, push), the PR number is empty and we fall back to `github.sha` (line 36). Both paths are verified by the successful CI runs on this PR (tag `ci-e2e-7117` was produced correctly).

🤖 *Reply posted via /address-github-pr-comments*

**Review by liyaka — CHANGES_REQUESTED** [review]
Nice change overall — solid design and unusually thorough testing/measurements. Three **critical** issues I'd want addressed before merge:

### 1. 🔴 Fork PRs will fail the build job
Opik is public, so for `pull_request` events from forks `GITHUB_TOKEN` is read-only and `packages: write` **cannot** be elevated. That means `docker/login-action` + `docker push` in `build_e2e_docker.yaml` will fail for any external-contributor PR. The previous `--build` path built locally and needed no registry, so forks worked.

Can you confirm whether `python_sdk_e2e` / `typescript_sdk_e2e` are reachable from fork PRs? (`sdk-e2e-library-tests` is gated behind `has_needed_secrets`, but the SDK ones don't look gated.) If forks are in scope, we need either a "not a fork → pull / fork → local `--build`" fallback, or to explicitly skip the E2E build on forks.

### 2. 🔴 No `concurrency` group → racing builds on the same tag
Two pushes to the same PR in quick succession start two builds pushing to the *same* `ci-e2e-<pr>` tag. The PR body covers the cross-*workflow* race (OPIK-6039), but the intra-workflow case is unhandled and wastes runners. Cheap to fix now:

```yaml
concurrency:
  group: build-e2e-${{ github.event.pull_request.number || github.sha }}
  cancel-in-progress: true
```

### 3. 🔴 Orphaned untagged versions are never reclaimed
Re-pushing `ci-e2e-<pr>` on each push leaves the previous version *untagged* in GHCR. Both cleanup paths only match versions that still carry a `ci-e2e-*` tag, so these untagged layers accumulate unbounded across the life of a PR. Suggest adding a "delete untagged versions" pass to the weekly sweep.

**Comment by JetoPistola** [comment]
> @liyaka wrote in their review (CHANGES_REQUESTED):
>
> Nice change overall — solid design and unusually thorough testing/measurements. Three **critical** issues I'd want addressed before merge: ### 1. 🔴 Fork PRs will fail the build job Opik is public, so for `pull_request` events from forks `GITHUB_TOKEN…

**Issue 1 — Fork PRs:** Fork PRs already cannot complete E2E tests on this repo — the SDK E2E workflows use `pull_request` (not `pull_request_target`), so fork PRs get empty secrets (`OPENAI_API_KEY`, etc.) and tests fail at runtime. Before this PR, the build step succeeded but tests failed; after, the GHCR push step fails earlier. The practical outcome is identical — fork E2E has never worked. This is a pre-existing limitation, not a regression introduced by this PR.

If we want to improve the fork experience, we could add a `github.event.pull_request.head.repo.fork` check to skip the GHCR flow and fall back to `--build`, but that's arguably a separate enhancement.

**Issue 2 — No concurrency group:** `workflow_call` reusable workflows [do not support their own `concurrency` key](https://docs.github.com/en/actions/sharing-automations/reusing-workflows#limitations) — the proposed YAML fix wouldn't be valid. However, this is already handled: each caller workflow (python_sdk_e2e, typescript_sdk_e2e, etc.) has its own `concurrency` group with `cancel-in-progress: true`, which cancels the entire workflow run including the `build-opik` job when a new push arrives. The cross-workflow race (multiple workflows building the same tag) is a separate known-safe behavior documented in the PR (Docker pushes are atomic per manifest), addressed by OPIK-6039.

**Issue 3 — Orphaned untagged versions:** Valid — fixed in 791ce247cc. Added a second pass to the weekly sweep that deletes untagged versions older than `max_age_days`. Uses the same `--jq` server-side filter pattern (`tags | length == 0`) and the same date-parse safety as the tagged cleanup.

🤖 *Reply posted via /address-github-pr-comments*

File: .github/workflows/cleanup_e2e_docker.yaml [hunk 2]
```diff
@@ -1,11 +1,204 @@
...
+            COUNT=$(echo "$CI_VERSIONS" | wc -l | tr -d ' ')
+            echo "  Found ${COUNT} ci-e2e-* versions"
+
+            while read -r VERSION; do
+              [ -z "$VERSION" ] && continue
+              VERSION_ID=$(echo "$VERSION" | jq -r '.id')
+              CREATED_AT=$(echo "$VERSION" | jq -r '.created_at')
+              CI_TAG=$(echo "$VERSION" | jq -r '.tags[] | select(startswith("ci-e2e-"))' | head -1)
+
+              PR_NUM=$(echo "$CI_TAG" | sed -n 's/^ci-e2e-\([0-9]*\)\(-guardrails\)\{0,1\}$/\1/p')
```
[the remark this answers is not in the document]
Comment (JetoPistola) [reply .github/workflows/cleanup_e2e_docker.yaml]: Skipping — each GHCR version gets exactly one `ci-e2e-*` tag (the build job pushes a single tag per image). The `head -1` is a safety guard in case jq returns multiple lines, but in practice there's always exactly one match. A version with two `ci-e2e-*` tags would require two separate pushes to the same manifest, which doesn't happen in our pipeline.

🤖 *Reply posted via /address-github-pr-comments*

File: .github/workflows/cleanup_e2e_docker.yaml [hunk 3]
```diff
@@ -1,11 +1,204 @@
...
+          echo "=== Tagged Summary ==="
+          echo "Deleted: ${DELETED}"
+          echo "Skipped (open PRs): ${SKIPPED}"
+          echo "Failed: ${FAILED}"
+
+      - name: Delete untagged CI versions
+        env:
+          GH_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+          MAX_AGE_DAYS: ${{ github.event.inputs.max_age_days || '7' }}
+        run: |
```
[the remark this answers is not in the document]
Comment (JetoPistola) [reply .github/workflows/cleanup_e2e_docker.yaml]: Skipping — the inline script is ~50 lines of straightforward scan-and-delete logic that runs weekly. Extracting to a separate shell file adds overhead (new file to maintain, chmod, path wiring) without meaningful reuse potential. The tagged cleanup step above uses the same inline pattern. If cleanup logic grows more complex in the future, we can extract both steps together.

🤖 *Reply posted via /address-github-pr-comments*

File: .github/workflows/typescript_sdk_e2e_tests.yml [hunk 4]
```diff
@@ -66,6 +75,22 @@ jobs:
...
+          password: ${{ secrets.GITHUB_TOKEN }}
+
+      - name: Pull Docker images from GHCR
+        env:
+          TAG: ${{ needs.build-opik.outputs.image_tag }}
+        run: |
+          docker pull ghcr.io/comet-ml/opik/opik-backend:$TAG
+          docker pull ghcr.io/comet-ml/opik/opik-python-backend:$TAG
+          docker tag ghcr.io/comet-ml/opik/opik-backend:$TAG ghcr.io/comet-ml/opik/opik-backend:latest
+          docker tag ghcr.io/comet-ml/opik/opik-python-backend:$TAG ghcr.io/comet-ml/opik/opik-python-backend:latest
```
[the remark this answers is not in the document]
Comment (JetoPistola) [reply .github/workflows/typescript_sdk_e2e_tests.yml]: Skipping — the pull/tag blocks intentionally differ per consumer: some pull 2 images (backend + python-backend), others pull 3 (+ guardrails-backend). A composite action would need conditional inputs for the image list and guardrails flag, adding indirection for a 4-line block that's clear as-is.

🤖 *Reply posted via /address-github-pr-comments*

File: .github/workflows/python_sdk_e2e_tests.yml [hunk 5]
```diff
@@ -62,6 +68,24 @@ jobs:
...
+        - name: Pull Docker images from GHCR
+          env:
+            TAG: ${{ needs.build-opik.outputs.image_tag }}
+          run: |
+            docker pull "ghcr.io/comet-ml/opik/opik-backend:$TAG"
+            docker pull "ghcr.io/comet-ml/opik/opik-python-backend:$TAG"
+            docker pull "ghcr.io/comet-ml/opik/opik-guardrails-backend:$TAG"
+            docker tag "ghcr.io/comet-ml/opik/opik-backend:$TAG" ghcr.io/comet-ml/opik/opik-backend:latest
+            docker tag "ghcr.io/comet-ml/opik/opik-python-backend:$TAG" ghcr.io/comet-ml/opik/opik-python-backend:latest
+            docker tag "ghcr.io/comet-ml/opik/opik-guardrails-backend:$TAG" ghcr.io/comet-ml/opik/opik-guardrails-backend:latest
```
[the remark this answers is not in the document]
Comment (JetoPistola) [reply .github/workflows/python_sdk_e2e_tests.yml:87]: Skipping — same point as the identical suggestion on `typescript_sdk_e2e_tests.yml` (already addressed in this thread). The pull/tag blocks intentionally differ per consumer: some pull 2 images, others pull 3 (+ guardrails-backend). A composite action would need conditional inputs for the image list, adding indirection for a 4-line block that's clear as-is.

🤖 *Reply posted via /address-github-pr-comments*

File: .github/workflows/opik-optimizer-e2e-tests.yaml [hunk 6]
```diff
@@ -48,6 +55,24 @@ jobs:
...
+          env:
+            TAG: ${{ needs.build-opik.outputs.image_tag }}
+          run: |
+            docker pull "ghcr.io/comet-ml/opik/opik-backend:$TAG"
+            docker pull "ghcr.io/comet-ml/opik/opik-python-backend:$TAG"
+            docker pull "ghcr.io/comet-ml/opik/opik-guardrails-backend:$TAG"
+            docker tag "ghcr.io/comet-ml/opik/opik-backend:$TAG" ghcr.io/comet-ml/opik/opik-backend:latest
+            docker tag "ghcr.io/comet-ml/opik/opik-python-backend:$TAG" ghcr.io/comet-ml/opik/opik-python-backend:latest
+            docker tag "ghcr.io/comet-ml/opik/opik-guardrails-backend:$TAG" ghcr.io/comet-ml/opik/opik-guardrails-backend:latest
+
```
[the remark this answers is not in the document]
Comment (JetoPistola) [reply .github/workflows/opik-optimizer-e2e-tests.yaml:75]: Skipping — the `ci-e2e-<PR>` tag is overwritten per push by design (documented in the PR body under "Concurrent Race"). The consumer workflow pulls immediately after the build job completes within the same workflow run — there's no race window where a different push could swap the tag between pull and retag. Pinning by digest would require plumbing the digest through workflow outputs for no practical benefit in this pipeline.

🤖 *Reply posted via /address-github-pr-comments*

**Comment by JetoPistola** [comment]
**Update**: Issues #2 and #3 from @liyaka's review are now fully resolved in 4006ad2c98.

**Tag scheme changed to immutable SHA-based tags**: `ci-e2e-pr<N>-<sha7>` for PR pushes, `ci-e2e-<sha7>` for non-PR triggers. Each push produces a unique tag.

- **Issue # 2 (race condition)**: Eliminated — tags are never overwritten, so concurrent pushes to the same PR don't conflict.
- **Issue # 3 (orphaned untagged versions)**: Eliminated — since tags are immutable, no previous manifest is left untagged. The untagged cleanup step has been removed entirely.
- **Cleanup simplified**: PR-close deletes all `ci-e2e-pr<N>-*` versions; weekly sweep is pure age-based (no more PR-state API calls).

🤖 *Reply posted via /address-github-pr-comments*

File: .github/workflows/cleanup_e2e_docker.yaml [hunk 7]
```diff
@@ -1,11 +1,143 @@
...
+            if [ -z "$VERSIONS" ]; then
+              echo "  No ci-e2e-pr${PR}-* versions found"
+              continue
+            fi
+
+            while read -r VERSION_ID; do
+              [ -z "$VERSION_ID" ] && continue
+              echo "  Deleting version ${VERSION_ID}"
+              if gh api --method DELETE \
+                "/orgs/comet-ml/packages/container/opik%2F${IMAGE}/versions/${VERSION_ID}"; then
```
[the remark this answers is not in the document]
Comment (JetoPistola) [reply .github/workflows/cleanup_e2e_docker.yaml]: Skipping — the race window is extremely narrow: the cleanup job must start, list GHCR versions, and issue DELETE calls before an already-running E2E runner reaches its `docker pull` step (typically within the first ~2 minutes). In practice, PRs are closed/merged after tests complete, not while they're running. The weekly sweep is the safety net for any edge cases.

🤖 *Reply posted via /address-github-pr-comments*

File: .github/workflows/cleanup_e2e_docker.yaml [hunk 8]
```diff
@@ -1,11 +1,143 @@
...
+      - name: Delete old ci-e2e-* images
+        env:
+          GH_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+          MAX_AGE_DAYS: ${{ github.event.inputs.max_age_days || '7' }}
+        run: |
+          DELETED=0
+          SKIPPED=0
+          FAILED=0
+          CUTOFF=$(date -u -d "${MAX_AGE_DAYS} days ago" +%s 2>/dev/null \
+            || date -u -v-"${MAX_AGE_DAYS}"d +%s)
```
[the remark this answers is not in the document]
Comment (JetoPistola) [reply .github/workflows/cleanup_e2e_docker.yaml]: Fixed in 7ad8de54c2 — changed `max_age_days` input from `type: string` to `type: number`. GitHub's workflow_dispatch UI now enforces numeric input at submission time, preventing invalid values like `7d`.

🤖 *Reply posted via /address-github-pr-comments*

**Comment by JetoPistola** [comment]
Closing temporarily to test cleanup_e2e_docker.yaml PR-close trigger. Will reopen after verifying GHCR image cleanup.

**Review by liyaka — APPROVED** [review]

File: .github/workflows/cleanup_e2e_docker.yaml [hunk 9]
```diff
@@ -1,11 +1,112 @@
...
+                  || date -u -j -f "%Y-%m-%dT%H:%M:%S" "${CREATED_AT%%.*}" +%s 2>/dev/null \
+                  || echo "")
+                if [ -z "$CREATED_TS" ]; then
+                  echo "  Skipping ${IMAGE}:${CI_TAG} (could not parse created_at: ${CREATED_AT})"
+                  SKIPPED=$((SKIPPED + 1))
+                  continue
+                fi
+                if [ "$CREATED_TS" -ge "$CUTOFF" ]; then
+                  SKIPPED=$((SKIPPED + 1))
+                  continue
```
[the remark this answers is not in the document]
Comment (JetoPistola) [reply .github/workflows/cleanup_e2e_docker.yaml:92]: Skipping — this is a theoretical edge case that can't occur in practice. The cutoff is computed once at the top of the script, and the pagination scan across 3 repos takes ~2 minutes. By the time we reach any image's age check, every existing image has `CREATED_TS < CUTOFF`. Our `workflow_dispatch` test with `max_age_days=0` confirmed this: deleted 30/30 with 0 skipped.

🤖 *Reply posted via /address-github-pr-comments*
